File: rhizome/tools/entries.py
"""CRUD + search operations for KnowledgeEntry objects."""

from sqlalchemy import select
from sqlalchemy.ext.asyncio import AsyncSession

from rhizome.db import CurriculumTopic, KnowledgeEntry, Topic
from rhizome.db.models import EntryType


async def create_entry(
    session: AsyncSession,
    *,
    topic_id: int,
    title: str,
    content: str,
    entry_type: EntryType | None = None,
    additional_notes: str = "",
    difficulty: int | None = None,
    speed_testable: bool = False,
) -> KnowledgeEntry:
    """Create a new knowledge entry under a topic."""
    entry = KnowledgeEntry(
        topic_id=topic_id,
        title=title,
        content=content,
        entry_type=entry_type,
        additional_notes=additional_notes,
        difficulty=difficulty,
        speed_testable=speed_testable,
    )
    session.add(entry)
    await session.flush()
    return entry


async def get_entry(
    session: AsyncSession,
    entry_id: int,
) -> KnowledgeEntry | None:
    """Return an entry by id, or None if not found."""
    return await session.get(KnowledgeEntry, entry_id)


async def list_entries(
    session: AsyncSession,
    topic_id: int,
) -> list[KnowledgeEntry]:
    """Return all entries for a topic, ordered by created_at."""
    result = await session.execute(
        select(KnowledgeEntry)
        .where(KnowledgeEntry.topic_id == topic_id)
        .order_by(KnowledgeEntry.created_at)
    )
    return list(result.scalars().all())


async def update_entry(
    session: AsyncSession,
    entry_id: int,
    *,
    title: str | None = None,
    content: str | None = None,
    entry_type: EntryType | None = None,
    additional_notes: str | None = None,
    difficulty: int | None = None,
    speed_testable: bool | None = None,
) -> KnowledgeEntry:
    """Update an entry's fields. Only provided (non-None) fields are changed."""
    entry = await session.get(KnowledgeEntry, entry_id)
    if entry is None:
        raise ValueError(f"KnowledgeEntry {entry_id} not found")
    if title is not None:
        entry.title = title
    if content is not None:
        entry.content = content
    if entry_type is not None:
        entry.entry_type = entry_type
    if additional_notes is not None:
        entry.additional_notes = additional_notes
    if difficulty is not None:
        entry.difficulty = difficulty
    if speed_testable is not None:
        entry.speed_testable = speed_testable
    await session.flush()
    return entry


async def delete_entry(
    session: AsyncSession,
    entry_id: int,
) -> None:
    """Delete a knowledge entry."""
    entry = await session.get(KnowledgeEntry, entry_id)
    if entry is None:
        raise ValueError(f"KnowledgeEntry {entry_id} not found")
    await session.delete(entry)
    await session.flush()


async def search_entries(
    session: AsyncSession,
    query: str,
    *,
    topic_id: int | None = None,
    curriculum_id: int | None = None,
) -> list[KnowledgeEntry]:
    """Search entries by LIKE on title + content.

    Optionally scope to a specific topic or curriculum.
    """
    pattern = f"%{query}%"
    stmt = select(KnowledgeEntry).where(
        (KnowledgeEntry.title.ilike(pattern)) | (KnowledgeEntry.content.ilike(pattern))
    )
    if topic_id is not None:
        stmt = stmt.where(KnowledgeEntry.topic_id == topic_id)
    if curriculum_id is not None:
        stmt = (
            stmt
            .join(Topic, KnowledgeEntry.topic_id == Topic.id)
            .join(CurriculumTopic, Topic.id == CurriculumTopic.topic_id)
            .where(CurriculumTopic.curriculum_id == curriculum_id)
        )
    result = await session.execute(stmt)
    return list(result.scalars().all())

File: rhizome-rs/src/db/entries.rs
//! CRUD + Search operations for knowledge entries

use sqlx::SqlitePool;
use crate::db::models::{ EntryType, KnowledgeEntry };

#[derive(Debug, thiserror::Error)]
pub enum EntryError {
    #[error("KnowledgeEntry {0} not found")]
    NotFound(i64),

    #[error(transparent)]
    Db(#[from] sqlx::Error),
}

pub async fn create_entry(
    pool: &SqlitePool,
    topic_id: i64,
    title: &str,
    content: &str,
    entry_type: Option<EntryType>,
    additional_notes: Option<&str>,
    difficulty: Option<i64>,
    speed_testable: bool,
) -> Result<KnowledgeEntry, EntryError> {
    let notes = additional_notes.unwrap_or("");
    let entry = sqlx::query_as::<_, KnowledgeEntry>(
        "INSERT INTO knowledge_entry
             (topic_id, title, content, additional_notes,
              entry_type, difficulty, speed_testable)
         VALUES (?, ?, ?, ?, ?, ?, ?)
         RETURNING *",
    )
    .bind(topic_id)
    .bind(title)
    .bind(content)
    .bind(notes)
    .bind(entry_type)
    .bind(difficulty)
    .bind(speed_testable)
    .fetch_one(pool)
    .await?;
    Ok(entry)
}

pub async fn get_entry(
    pool: &SqlitePool,
    entry_id: i64,
) -> Result<Option<KnowledgeEntry>, EntryError> {
    let entry = sqlx::query_as::<_, KnowledgeEntry>(
        "SELECT * FROM knowledge_entry WHERE id = ?",
    )
    .bind(entry_id)
    .fetch_optional(pool)
    .await?;
    Ok(entry)
}

pub async fn count_entries(
    pool: &SqlitePool,
    topic_id: i64,
) -> Result<i64, EntryError> {
    let row: (i64,) = sqlx::query_as(
        "SELECT COUNT(*) FROM knowledge_entry WHERE topic_id = ?",
    )
    .bind(topic_id)
    .fetch_one(pool)
    .await?;
    Ok(row.0)
}

pub async fn list_entries(
    pool: &SqlitePool,
    topic_id: i64,
) -> Result<Vec<KnowledgeEntry>, EntryError> {
    let entries = sqlx::query_as::<_, KnowledgeEntry>(
        "SELECT * FROM knowledge_entry
          WHERE topic_id = ?
          ORDER BY created_at",
    )
    .bind(topic_id)
    .fetch_all(pool)
    .await?;
    Ok(entries)
}

pub async fn update_entry(
    pool: &SqlitePool,
    entry_id: i64,
    title: Option<&str>,
    content: Option<&str>,
    entry_type: Option<EntryType>,
    additional_notes: Option<&str>,
    difficulty: Option<i64>,
    speed_testable: Option<bool>,
) -> Result<KnowledgeEntry, EntryError> {
    let existing = get_entry(pool, entry_id)
        .await?
        .ok_or(EntryError::NotFound(entry_id))?;
    let final_title = title.unwrap_or(&existing.title);
    let final_content = content.unwrap_or(&existing.content);
    let final_entry_type = entry_type.or(existing.entry_type);
    let final_notes = additional_notes.unwrap_or(&existing.additional_notes);
    let final_difficulty = difficulty.or(existing.difficulty);
    let final_speed_testable = speed_testable.unwrap_or(existing.speed_testable);
    let entry = sqlx::query_as::<_, KnowledgeEntry>(
        "UPDATE knowledge_entry
            SET title = ?, content = ?, entry_type = ?,
                additional_notes = ?, difficulty = ?,
                speed_testable = ?, updated_at = unixepoch()
          WHERE id = ?
         RETURNING *",
    )
    .bind(final_title)
    .bind(final_content)
    .bind(final_entry_type)
    .bind(final_notes)
    .bind(final_difficulty)
    .bind(final_speed_testable)
    .bind(entry_id)
    .fetch_one(pool)
    .await?;
    Ok(entry)
}

pub async fn delete_entry(
    pool: &SqlitePool,
    entry_id: i64,
) -> Result<(), EntryError> {
    let result = sqlx::query("DELETE FROM knowledge_entry WHERE id = ?")
        .bind(entry_id)
        .execute(pool)
        .await?;
    if result.rows_affected() == 0 {
        return Err(EntryError::NotFound(entry_id));
    }
    Ok(())
}

pub async fn search_entries(
    pool: &SqlitePool,
    query: &str,
    topic_id: Option<i64>,
) -> Result<Vec<KnowledgeEntry>, EntryError> {
    let pattern = format!("%{}%", query);
    let entries = sqlx::query_as::<_, KnowledgeEntry>(
        "SELECT * FROM knowledge_entry
          WHERE (title LIKE ?1 OR content LIKE ?1)
            AND (?2 IS NULL OR topic_id = ?2)",
    )
    .bind(&pattern)
    .bind(topic_id)
    .fetch_all(pool)
    .await?;
    Ok(entries)
}

